engine/storage/src/main/java/org/apache/cloudstack/storage/endpoint/DefaultEndPointSelector.java

```diff
@@ -115,22 +123,46 @@ protected EndPoint findEndPointInScope(Scope scope, String sqlBase, Long poolId)
         StringBuilder sbuilder = new StringBuilder();
         sbuilder.append(sqlBase);
 
+        List<Long> dedicatedHosts = new ArrayList<Long>();
+
         if (scope != null) {
             if (scope.getScopeType() == ScopeType.HOST) {
                 sbuilder.append(" and h.id = ");
                 sbuilder.append(scope.getScopeId());
             } else if (scope.getScopeType() == ScopeType.CLUSTER) {
                 sbuilder.append(" and h.cluster_id = ");
                 sbuilder.append(scope.getScopeId());
+                DedicatedResourceVO dedicatedHost = dedicatedResourceDao.findByClusterId(scope.getScopeId());
+                if (dedicatedHost != null){
+                    List<HostVO> clusterHosts = hostDao.findByClusterId(scope.getScopeId());
+                    for (HostVO hostVO: clusterHosts){
+                        dedicatedHosts.add(hostVO.getId());
+                    }
+                }
```

Comment:
@Spaceman1984 what if cluster is not dedicated, but some hosts in that cluster are dedicated?

Reply:
This is handled on line 155